cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStoreProvider.java

```diff
@@ -0,0 +1,543 @@
+    FileSecureStoreProvider fileSecureStoreProvider = new FileSecureStoreProvider(cConf);
+    // Locate and load the keystore
+    fileSecureStoreProvider.locateKeystore();
+    return fileSecureStoreProvider;
+  }
+
+  /**
+   * Stores an element in the secure store. If the key already exists then delete it first.
+   * @param name
+   * @param data
+   * @param properties
+   * @throws IOException
+   */
+  void put(String name, byte[] data, Map<String, String> properties) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    writeLock.lock();
+    try {
+      if (keyStore.containsAlias(name) || cache.containsKey(metaKey)) {
+        // Clear the existing key so that we can write the new one.
+        delete(name);
+      }
+      keyStore.setKeyEntry(name, new SecretKeySpec(data, ALGORITHM_PROXY),
+                           password, null);
+      SecureStoreMetadata meta = FileSecureStoreMetadata.of(name, properties);
+      /*
+        The data was written successfully to the key store. Now try to write the metddata.
+        If this fails then try to delete the data and throw an exception.
+       */
+      try {
+        keyStore.setKeyEntry(metaKey, new KeyMetadata(meta), password, null);
+      } catch (KeyStoreException k) {
+        keyStore.deleteEntry(name);
+        throw k;
+      }
+      cache.put(metaKey, meta);
+    } catch (KeyStoreException e) {
+      throw new IOException("Failed to store the key. ", e);
+    } finally {
+      writeLock.unlock();
+    }
+    // Attempt to persist the store.
+    flush();
+  }
+
+  /**
+   * Deletes the element with the given name.
+   * @param name
+   */
+  void delete(String name) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    writeLock.lock();
+    if (cache.containsKey(metaKey)) {
+      cache.remove(metaKey);
+    }
+    try {
+      if (keyStore.containsAlias(name)) {
+        keyStore.deleteEntry(name);
+      }
+      /* It's OK if we were able to delete the data but not the metadata.
+         Java keystore allows overwrites, so the next time we try to write a
+         key with the same name the metadata will be overwritten. */
+      if (keyStore.containsAlias(metaKey)) {
+        keyStore.deleteEntry(metaKey);
+      }
+    } catch (KeyStoreException e) {
+      LOG.error("Failed to delete the key " + name, e);
+    } finally {
+      writeLock.unlock();
+    }
+    // Attempt to persist the store.
+    flush();
+  }
+
+  /**
+   * List of all the entries in the secure store.
+   * @return A map of name -> description
+   * @throws IOException
+   */
+  Map<String, String> list() throws IOException {
+    Map<String, String> list = new HashMap<>();
+    String name = null;
+    readLock.lock();
+    try {
+      try {
+        Enumeration<String> aliases = keyStore.aliases();
+        // Surprisingly enhanced for-each does not work for Enumeration
```

Comment:
I think that's because you can only go through an `Enumeration` once.
It has to be Iterable to use the enhanced for-each